### PR TITLE
remove_asm now guarantees that functions called exist

### PR DESCRIPTION
**Pull request by kroening**
This fixes the problem that remove_asm introduces calls to functions that subsequently don't exist in the function_map.

Consequently, it is now no longer required to do goto_conversion after remove_asm(), greatly simplifying the use of remove_asm().

### CODE REVIEW

**Comment by smowton** [comment]
I presume it is no coincidence this appears concurrently with @polgreen's #2636

File: src/goto-programs/remove_asm.cpp
Comment (smowton) [inline src/goto-programs/remove_asm.cpp]: insert / emplace?

**Review by smowton — APPROVED** [review]

**Review by tautschnig — CHANGES_REQUESTED** [review]
It seems this might fix #2631; if so, a test needs to be added. The GitHub PR comment is very helpful (and indicates that a bug is fixed), but unfortunately the commit message lacks all such information. Please fix.

**Comment by polgreen** [comment]
This is going to have the same problem as I have in my PR as soon as you introduce tests:
- windows CI won't deal with `__asm__ __volatile__ ("mfence": : :"memory");`
- CBMC doesn't treat `__asm mfence` in the same way as `__asm__ __volatile__ ("mfence": : :"memory");`. 

As a result you won't be able to have any tests without including a hack like I have done in my PR. 

I also don't think it is the right fix to force every function to be added to the function_map. All other places in CBMC, where the function map is accessed, if the function is not found in the map it is treated as having no body.

**Comment by kroening** [comment]
I'll re-use the test in #2631 once that's merged. This PR will eliminate the need for the 'windows hack'.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: b8c86e737b83f42b9440656c59297ea5f89cf8af).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80492971

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: d81a7130651e506b23b7ba5995918481ea556ae6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80494821
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Comment by kroening** [comment]
This now removes the need for the 'hack' in #2636

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 21feec41ec991a6fff56624f25747cc83a121820).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80498081
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 45141d428273515e2b0aaa97d25c685c4122f7e9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80502428

**Comment by polgreen** [comment]
I would be very keen on this being merged. It appears that assembly functions not appearing in the function map causes a multitude of issues all over the place.

File: src/goto-programs/remove_asm.cpp
Comment (tautschnig) [reply src/goto-programs/remove_asm.cpp]: +1 on this suggestion. This may require adding a constructor to `goto_functiont`.

File: src/goto-programs/remove_asm.cpp
Comment (tautschnig) [inline src/goto-programs/remove_asm.cpp]: Nit picking, but should this change possibly go in the previous commit?

File: src/goto-programs/remove_asm.cpp
Comment (tautschnig) [inline src/goto-programs/remove_asm.cpp]: This is an important bugfix that may warrant its own commit (and regression test - @polgreen may have one).

File: src/goto-programs/remove_asm.cpp
Comment (tautschnig) [inline src/goto-programs/remove_asm.cpp]: I'd rather see `UNIMPLEMENTED` or the likes here. Those silent failures have bitten us before.

**Review by tautschnig — APPROVED** [review]
Approving with a couple of requests that I'd really like to see addressed...

File: src/goto-programs/remove_asm.cpp
Comment (kroening) [reply src/goto-programs/remove_asm.cpp]: Moved to PR #2714

File: src/goto-programs/remove_asm.cpp
Comment (kroening) [reply src/goto-programs/remove_asm.cpp]: Moved to PR #2711

File: src/goto-programs/remove_asm.cpp
Comment (kroening) [reply src/goto-programs/remove_asm.cpp]: Yes, will now happen after merging this, then rebasing #2714

**Comment by kroening** [comment]
On using emplace instead of if(find()==end()) { insert() }
I don't like this for two reasons:
1) Emplace doesn't actually guarantee that the constructor isn't run, which is the expensive thing here.
2) Emplace is much harder to read.
3) Any STL implementation I know uses finger search, i.e., the find ... insert idiom isn't slower than insert.

File: src/goto-programs/remove_asm.cpp
Comment (tautschnig) [inline src/goto-programs/remove_asm.cpp:115]: Nit picking: I'd still like to see using `.insert` or `.emplace` here instead of the redundancy of doing `.find` plus `.operator[]`.

**Review by tautschnig — APPROVED** [review]

File: src/goto-programs/remove_asm.cpp
Comment (tautschnig) [reply src/goto-programs/remove_asm.cpp:115]: Ah, sorry, I hadn't seen https://github.com/diffblue/cbmc/pull/2642#issuecomment-412100793.   In response to that: I believe `insert` will always do `find` + `insert-with-hint`, which ought to be faster. But indeed it's very minor.

**Pull request merged**
